modules/generator/processor/localblocks/query_range.go

```diff
@@ -0,0 +1,280 @@
+	overallEval.ObserveSeries(walResults)
+
+	return overallEval.Results(), nil
+}
+
+func (p *Processor) queryRangeWALBlock(ctx context.Context, b common.WALBlock, eval *traceql.MetricsEvalulator) error {
+	m := b.BlockMeta()
+	span, ctx := opentracing.StartSpanFromContext(ctx, "Processor.QueryRange.WALBlock", opentracing.Tags{
+		"block":     m.BlockID,
+		"blockSize": m.Size,
+	})
+	defer span.Finish()
+
+	fetcher := traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+		return b.Fetch(ctx, req, common.DefaultSearchOptions())
+	})
+
+	return eval.Do(ctx, fetcher, uint64(m.StartTime.UnixNano()), uint64(m.EndTime.UnixNano()))
+}
+
+func (p *Processor) queryRangeCompleteBlock(ctx context.Context, b *ingester.LocalBlock, req tempopb.QueryRangeRequest, timeOverlapCutoff float64, unsafe bool) ([]*tempopb.TimeSeries, error) {
+	m := b.BlockMeta()
+	span, ctx := opentracing.StartSpanFromContext(ctx, "Processor.QueryRange.CompleteBlock", opentracing.Tags{
+		"block":     m.BlockID,
+		"blockSize": m.Size,
+	})
+	defer span.Finish()
+
+	// Trim and align the request for this block. I.e. if the request is "Last Hour" we don't want to
+	// cache the response for that, we want only the few minutes time range for this block. This has
+	// size savings but the main thing is that the response is reuseable for any overlapping query.
+	req.Start, req.End = traceql.TrimToOverlap(req.Start, req.End, req.Step, uint64(m.StartTime.UnixNano()), uint64(m.EndTime.UnixNano()))
+
+	if req.Start >= req.End {
+		// After alignment there is no overlap or something else isn't right
+		return nil, nil
+	}
+
+	cached, name, err := p.queryRangeCacheGet(ctx, m, req)
+	if err != nil {
+		return nil, err
+	}
+
+	span.SetTag("cached", cached != nil)
+
+	if cached != nil {
+		return cached.Series, nil
+	}
+
+	// Not in cache or not cacheable, so execute
+	eval, err := traceql.NewEngine().CompileMetricsQueryRange(&req, false, timeOverlapCutoff, unsafe)
+	if err != nil {
+		return nil, err
+	}
+	f := traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+		return b.Fetch(ctx, req, common.DefaultSearchOptions())
+	})
+	err = eval.Do(ctx, f, uint64(m.StartTime.UnixNano()), uint64(m.EndTime.UnixNano()))
+	if err != nil {
+		return nil, err
+	}
+
+	results := eval.Results().ToProto(&req)
+
+	if name != "" {
+		err = p.queryRangeCacheSet(ctx, m, name, &tempopb.QueryRangeResponse{
+			Series: results,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("writing local query cache: %w", err)
+		}
+	}
+
+	return results, nil
+}
+
+func (p *Processor) queryRangeCacheGet(ctx context.Context, m *backend.BlockMeta, req tempopb.QueryRangeRequest) (*tempopb.QueryRangeResponse, string, error) {
+	cacheable := req.Start <= uint64(m.StartTime.UnixNano()) && req.End >= uint64(m.EndTime.UnixNano())
+	if !cacheable {
+		return nil, "", nil
+	}
```

Comment:
Agree, I think the requirement for caching changed during implementation.  Originally it was trying to avoid partial time range caching, but after solidifying the time handling with TrimToOverlap, I missed this. Will take a look.

Reply:
Cleaned up